templates/theories/linear_algebra.md

```diff
@@ -0,0 +1,106 @@
+An additive commutative monoid `M` is a semimodule over the semiring `R` if there is a scalar multiplication `•` (`has_scalar.smul`) that satisfies the expected distributivity axioms for `+` (in `M` and `R`) and `*` (in `R`).
+A module (typeclass `module`) is a semimodule that additionally requires that `R` is a ring and `M` is a group.
+A vector space (typeclass `vector_space`) is a module that additionally requires that `R` is a field.
+All vector spaces are also modules, and all modules are also semimodules.
+
+Typical examples are that `m → ℕ` is an `ℕ`-semimodule, `m → ℤ` is a `ℤ`-module and `m → ℚ` is a `ℚ`-vector space
+(outside of type theory, these are known as `ℕ^m`, `ℤ^m` and `ℚ^m` respectively).
+These instances are defined in `algebra.pi_instances`.
+A (semi)ring is a (semi)module over itself, with `•` defined as `*` (`smul_eq_mul`).
+
+The file `linear_algebra.basis` defines linear independence and bases for modules.
+
+The file `linear_algebra.dimension` defines the dimension of a vector space as the minimum cardinality of a basis.
+Tre rank of a linear map is defined as the dimension of its image.
+Most definitions in this file are non-computable.
+
+### matrices ###
+
+#### data.matrix.basic ###
+The type `matrix m n α` contains rectangular, `m` by `n` arrays of elements of the type `α`.
+It is an alias for the type `m → n → α`, under the assumptions `[fintype m] [fintype n]` stating that `m` and `n` have finitely many elements.
+A matrix type can be indexed over arbitrary `fintype`s.
+For example, the adjacency matrix of a graph could be indexed over the nodes in that graph.
+If you want to specify the dimensions of a matrix as natural numbers `m n : ℕ`, you can use `fin m` and `fin n` as index types.
+
+A matrix is constructed by giving the map from indices to entries: `(λ (i : m) (j : n), (_ : α)) : matrix m n α`.
+For matrices indexed by natural numbers, you can also use the notation defined in `data.matrix.notation`: `![![a, b, c], ![b, c, d]] : matrix (fin 2) (fin 3) α`.
+To get an entry of the matrix `M : matrix m n α` at row `i : m` and column `j : n`,
+you can apply `M` to the indices: `M i j : α`.
+Lemmas about the entries of a matrix typically end in `_val`: `add_val M N i j : (M + N) i j = M i j + N i j`.
+
+Matrix multiplication and transpose have notation that is made available by the command `open_locale matrix`.
+
+When working with matrices, a *vector* means a function `m → α` for an arbitrary `fintype` `m`.
+These have a module (or vector space) structure defined in `algebra.pi_instances`
+consisting of pointwise addition and multiplication.
+The distinction between row and column vectors is only made by the choice of function.
+For example, `mul_vec M v` multiplies a matrix with a column vector `v : m → α` and `vec_mul v M` multiplies a row vector `v : m → α` with a matrix.
```

Comment:
Are there notations for those operations?

Reply:
There aren't any. If you need a lot of `mul_vec` and `vec_mul`, I think you're supposed to use a `linear_map` instead anyway. I've added some text to that effect.